Delete / Publish / Unpublish actions on instructorHome redirects to Feedbacks page instead of Home. #3649

kennho opened the pull request:
Fixes #3649

kennho commented:
I was scrolling through the issues and I found #3360 which looks like a duplicate / subset of this? I also saw #3400 which is similar to this.

Actually, I have changed the "copy" and the "remind" (other than "publish / unpublish" and "delete") button on instructorHome to redirect back to home too. Should I remove the change and submit a pull request under #3400 instead?

kanghj commented on src/main/java/teammates/ui/controller/InstructorFeedbackDeleteAction.java:
```diff
@@ -29,7 +32,9 @@ protected ActionResult execute() throws EntityDoesNotExistException {
                         + "from Course: <span class=\"bold\">[" + courseId + " deleted.";
         
-        return createRedirectResult(Const.ActionURIs.INSTRUCTOR_FEEDBACKS_PAGE);
+        nextUrl = nextUrl == null ? Const.ActionURIs.INSTRUCTOR_FEEDBACKS_PAGE : nextUrl;
```
This seems to be repeated above?

kanghj commented on src/main/java/teammates/ui/controller/PageData.java:
```diff
@@ -525,11 +539,20 @@ public String getInstructorStudentCommentClearPendingLink(String courseId) {
+     * @param isHome true to indicate the request is from the home page
+     * @return the link to submit request to delete the session with return page link
+     */
+    public String getInstructorFeedbackDeleteLink(String courseId, String feedbackSessionName, boolean isHome) {
```
Why change from using <code>nextUrl</code> to <code>isHome</code>?

kennho replied:
oh because previously the buildBasicFsForm was passing an empty string inside. This is causing an error after I updated InstructorFeedbackDeleteAction to use nextUrl as nextUrl will give me something like `*nextUrl=user=text@example.com` 
Should I change it to take in String then pass in the correct String when calling instead?

kanghj replied:
Yes, I think it's better to pass in the correct String. We can have a default nextUrl if it's empty

kennho replied:
Okay. Do I do this for getInstructorFeedbackDeleteLink only or for the rest of the similar methods also? @kanghj

kanghj replied:
Let's do it for the similar methods that were modified in this pr.

kanghj commented:
@kennho, I think it would be better if this was done under separate issues, since this pr is quite big. But I'll leave the choice to you if you want to split this up into different issues or continue with what you already have here.

shreyamahapatra commented:
people..I have cloned the repository..followed all the step but could not access admin/adminHomePage.please help me out

thyageshm commented:
@shreyamahapatra please create a new issue... I will be removing your comment here because it is not relevant to this thread.

kennho commented:
Updated with suggested changes for delete, publish and unpublish link. Removed the fix to the "copy" button and submitted another PR on issue #3400 for the "copy" button. Since this PR is to solve the delete, publish and unpublish button, I am going to remove the fix to the "remind" button and create another issue on it. @kanghj

kennho commented:
Removed the fix to the "remind" button. Created issue #5088 and submitted a PR. Also updated methods to take in return url instead of taking in isHome. @kanghj

kanghj commented on src/main/java/teammates/ui/controller/InstructorHomeCourseAjaxPageData.java:
```diff
@@ -146,7 +146,7 @@ private void addAttributeIf(boolean shouldAdd, ElementTag button, String key, St
                     isRecent,
                     getInstructorFeedbackSessionActions(
-                            session, false, instructor));
+                            session, true, instructor));
```
Can you comment here why this change is made?

kennho replied:
oh. this is to set the isHome to true so the correct return url can be generated at the InstructorFeedbackSessionActions.java

kanghj commented on src/main/java/teammates/ui/controller/PageData.java:
```diff
@@ -525,12 +525,20 @@ public String getInstructorStudentCommentClearPendingLink(String courseId) {
+     * Retrieves the link to submit request to delete the session.
+     * @param courseId course ID
+     * @param feedbackSessionName the session name
+     * @param isHome true to indicate the request is from the home page
```
This javadoc parameter should be updated I guess

kanghj commented:
Let's wait for @yj-soh to give comments on #5089, since the three PRs you have are roughly similar

kennho commented:
Okay. Updated the comments too @kanghj

thyageshm commented:
@kanghj ready for review

yj-soh commented on src/test/java/teammates/test/cases/ui/browsertests/AdminHomePageUiTest.java:
```diff
@@ -237,23 +237,20 @@ private void testCreateInstructorAction() throws Exception {
         instructorHomePage.loadInstructorHomeTab();
-        feedbacksPage = instructorHomePage.clickFeedbackSessionRemindLink("AHPUiT.instr1.gma-demo", 
-                                                                          "Second team feedback session");
+        feedbacksPage = instructorHomePage.clickFeedbackSessionRemindLink("AHPUiT.instr1.gma-demo", "Second team feedback session");
```
Is there a reason for removing the line break for the sake of shortening the lines?

yj-soh commented on src/main/java/teammates/ui/template/FeedbackSessionPublishButton.java:
```diff
@@ -25,6 +25,8 @@ public FeedbackSessionPublishButton(PageData data, FeedbackSessionAttributes ses
         this.isSendingPublishedEmail = session.isPublishedEmailEnabled;
 
+        String returnUrl = isHome ? Const.ActionURIs.INSTRUCTOR_HOME_PAGE 
+                                  : Const.ActionURIs.INSTRUCTOR_FEEDBACKS_PAGE;
```
Is there a reason for passing in `isHome` as a parameter rather than `returnUrl`?

kennho replied:
opps sorry left out this part! will make the necessary changes

kennho commented:
Updated with suggested changes @yj-soh @kanghj

yj-soh commented on src/main/java/teammates/ui/template/InstructorFeedbackSessionActions.java:
```diff
@@ -34,6 +34,8 @@ public InstructorFeedbackSessionActions(PageData data, FeedbackSessionAttributes
         String courseId = session.courseId;
         String feedbackSessionName = session.feedbackSessionName;
+        String returnUrl = isHome ? Const.ActionURIs.INSTRUCTOR_HOME_PAGE
+                                  : Const.ActionURIs.INSTRUCTOR_FEEDBACKS_PAGE;
```
Actually, if we want extensibility, it's probably best to just use `returnUrl` from the start, with the `*PageData` calling such functions and passing in the URI rather than a boolean like `isHome`.

yj-soh commented on src/main/java/teammates/ui/template/InstructorFeedbackSessionActions.java:
```diff
@@ -61,8 +61,7 @@ public InstructorFeedbackSessionActions(PageData data, FeedbackSessionAttributes
             
-        this.publishButton = new FeedbackSessionPublishButton(data, session, isHome, instructor,
-                                                              PUBLISH_BUTTON_TYPE);
+        this.publishButton = new FeedbackSessionPublishButton(data, session, returnUrl, instructor,PUBLISH_BUTTON_TYPE);
```
Fix the formatting here.
Keep to our line length limits wherever possible.

yj-soh commented on src/test/java/teammates/test/cases/ui/browsertests/AdminHomePageUiTest.java:
```diff
@@ -237,23 +237,20 @@ private void testCreateInstructorAction() throws Exception {
+        instructorHomePage.verifyHtmlMainContent("/newlyJoinedInstructorFeedbackSessionPublished.html");
         
-        feedbacksPage.logout();
+        instructorHomePage.logout();
```
Does `feedbacksPage` need to be logged out?

kennho replied:
Hmm. I think they both calls the same method in AppData.java which is to click the logout button. Tried using either one. Both does the same thing of logging out.

kennho commented:
Updated with suggested changes @kanghj @yj-soh

yj-soh commented:
It looks okay to me, @kanghj, anything from you?

I do have some general comments on the PR, though.
When updating tests, why not put them in a single commit if they are related?
Having different commits for each file only makes it harder to trace the exact cause in the future and messes up the repo history.
The comment of "Updated test pages affected by {name}" is also pretty worthless when you don't actually mention why.

When merging in master to update the branch, try keeping that on your local repo and clean it up before pushing.
There's absolutely no need to have 10 consequent commits for merging in master.
Or rather, there's no need to even merge in master unless you require something new that is available only in master.
Those commits will end up in master if the branch gets merged, so all those unnecessary merge commits will end up polluting the master branch as well.

kennho commented:
Oops! Sorry! It is stated to sync with the committer repo frequently so I did that everytime there is an update to the master 😅  Will take note of it in the future. @yj-soh

damithc commented:
Get rid of noisy commits?

kennho commented:
do I still remove the multiple commits? I am currently trying to rebase and squash all the merge master commits into one and the update tests commits into one @kanghj

kanghj commented:
Yup, just do it

kennho commented:
Done removing @kanghj @yj-soh
